src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/ListenerPrimary.cs

```diff
@@ -179,5 +184,53 @@ private struct FILE_COMPLETION_INFORMATION
                 }, this).ConfigureAwait(false);
             }
         }
+
+        private class PipeReadContext
+        {
+            private readonly ListenerPrimary _listener;
+            private readonly IntPtr _bufPtr;
+            private int _bytesRead;
+
+            public PipeReadContext(ListenerPrimary listener)
+            {
+                _listener = listener;
+                _bufPtr = Marshal.AllocHGlobal(8);
```

Comment:
Allocating memory for a native struct VS allocating a buffer for reading feel like very different things. Sure they are different allocation patterns in the same class but for this it's not a big deal. Anyways, no real strong argument but using GC tracked memory is much nicer in general.